### PR TITLE
8284027: vmTestbase/nsk/jvmti/GetAllThreads/allthr001/ is failing

### PR DESCRIPTION
**Pull request by alexmenkov**
The test counts all "system" threads before the execution and expects that this number remains the same during test execution.
This makes the test fragile - JVM may start internal threads, some threads may end.

The fix updates the test:
- the test checks only test threads, and verify that the live threads are reported by GetAllThreads and terminated threads are not reported;
- dropped "system" thread counting stuff;
- added proper deallocation of GetThreadInfo results.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 reviews required, with at least 1 reviewer)

### Issue
 * [JDK-8284027](https://bugs.openjdk.java.net/browse/JDK-8284027): vmTestbase/nsk/jvmti/GetAllThreads/allthr001/ is failing

### Reviewers
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8512/head:pull/8512` \
`$ git checkout pull/8512`

Update a local copy of the PR: \
`$ git checkout pull/8512` \
`$ git pull https://git.openjdk.java.net/jdk pull/8512/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8512`

View PR using the GUI difftool: \
`$ git pr show -t 8512`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8512.diff">https://git.openjdk.java.net/jdk/pull/8512.diff</a>

</details>

### CODE REVIEW

**Review by sspitsyn — APPROVED** [review]
Alex,
It looks good to me.
Thank you for taking care about this!
Serguei

**Comment by sspitsyn** [comment]
Should we remove prefix "[LOOM] " from the bug title?

**Comment by alexmenkov** [comment]
> Should we remove prefix "[LOOM] " from the bug title?

Done

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp [hunk 1]
```diff
@@ -149,12 +167,23 @@ jint  Agent_Initialize(JavaVM *jvm, char *options, void *reserved) {
     return JNI_OK;
 }
 
+void release(JNIEnv *env, jvmtiThreadInfo *info) {
```
Comment (plummercj) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp:170]: How about "releaseThreadInfo"?

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp [hunk 2]
```diff
@@ -196,20 +225,33 @@ JNIEXPORT void checkInfo(JNIEnv *env, int ind) {
...
+        if (err != JVMTI_ERROR_NONE) {
+            printf("Failed to get thread info: %s (%d)\n",
+                   TranslateError(err), err);
+            result = STATUS_FAILED;
+            return;
+        }
+        if (printdump == JNI_TRUE) {
+            printf(" >>> %s", inf.name);
         }
+        for (j = 0, found = 0; j < thrInfo[ind].unexpected.cnt && !found; j++) {
```
Comment (plummercj) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp:240]: found = JNI_FALSE

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp [hunk 3]
```diff
@@ -196,20 +225,33 @@ JNIEXPORT void checkInfo(JNIEnv *env, int ind) {
...
-            num_unexpected++;
+        err = jvmti->GetThreadInfo(threads[i], &inf);
+        if (err != JVMTI_ERROR_NONE) {
+            printf("Failed to get thread info: %s (%d)\n",
+                   TranslateError(err), err);
+            result = STATUS_FAILED;
+            return;
+        }
+        if (printdump == JNI_TRUE) {
+            printf(" >>> %s", inf.name);
```
Comment (plummercj) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp:238]: Don't you want a newline here?

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp (hunk 1 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp (hunk 2 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp]: found is int, not jboolean.
I changed it to bool to be more C++-style

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp (hunk 3 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp:239]: No. This is pre-existing code.
The log looks like:
 >>> Check: 0
 >>> main >>> Reference Handler >>> Finalizer >>> Signal Dispatcher >>> Attach Listener >>> Notification Thread >>> Common-Cleaner >>> MainThread

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp [hunk 4]
```diff
@@ -149,12 +167,24 @@ jint  Agent_Initialize(JavaVM *jvm, char *options, void *reserved) {
...
+        env->DeleteLocalRef(info->context_class_loader);
+    }
+}
+
 JNIEXPORT void checkInfo(JNIEnv *env, int ind) {
     jint threadsCount = -1;
     jthread *threads;
-    int i, j, found;
+    int i, j;
+    bool found;
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp:184]: Nit: this isn't old style C code, we don't have to declare all variables up front. New variables should be declared where needed.

Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetAllThreads/allthr001/allthr001.cpp]: ok :)
fixed.

**Review by plummercj — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]
Update looks good.
Thanks,
Serguei

**Pull request closed without merging**